File: Week 4/templates/D_bridges.java
import java.io.*;
import java.util.*;

class Main{
    // Make sure when you submit, the class name is Main
	
    // Input format: adj.get(i) is a list of all the neighbors of node i
    //               is_single.get(i).get(j) tells whether the edge from i to adj.get(i).get(j) is a single-lane bridge
    //               adj.get(i).size() == is_single.get(i).size() for all i
    // The input is 1-indexed
    static int bridges(int n, ArrayList<ArrayList<Integer>> adj, ArrayList<ArrayList<Boolean>> is_single){
        // Your code goes here
        return -1; // This is a placeholder
    }

    public static void main(String[] args) throws Exception{
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(in.readLine());
        int n = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken());
        ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>();
        ArrayList<ArrayList<Boolean>> is_single = new ArrayList<ArrayList<Boolean>>();
        while(adj.size() < n + 1) {
            adj.add(new ArrayList<Integer>());
            is_single.add(new ArrayList<Boolean>());
        }
        for(int i = 0; i < m; i++){
            st = new StringTokenizer(in.readLine());
            int u = Integer.parseInt(st.nextToken()), v = Integer.parseInt(st.nextToken()), z = Integer.parseInt(st.nextToken());
            adj.get(u).add(v);
            is_single.get(u).add(z == 1);
            adj.get(v).add(u);
            is_single.get(v).add(z == 1);
        }
        System.out.println(bridges(n, adj, is_single));
    }
}

File: Week4/templates/D_bridges.cpp
#include <bits/stdc++.h>
using namespace std;

// Input format: adj[i] is a vector of the neighbors of node i
//               is_single[i][j] tells whether the edge from i to adj[i][j] is a single-lane bridge
//               adj[i].size() == is_single[i].size() for all i
// The input is 1-indexed
int bridges(int n, vector<vector<int>> adj, vector<vector<bool>> is_single){
    // In a later week we'll talk about generalized shortest paths problems
    // Dijkstra's algorithm solves this problem too 

    vector<int> dist(n + 1, INT_MAX);
    deque<int> dq; // The trick to this problem is to use a deque instead of a queue
                    // If a vertex isn't a bridge, we can push it to the front of the deque instead

    dq.push_back(1);
    dist[1] = 0;
    while(dq.size()){
        int u = dq.front();
        dq.pop_front();
        for(int i = 0; i < adj[u].size(); i++){
            int v = adj[u][i];
            int new_dist = dist[u] + (is_single[u][i] ? 1 : 0);
            if(new_dist < dist[v]){
                dist[v] = new_dist;
                if(new_dist == dist[u]){
                    dq.push_front(v);
                }else{
                    // It's fine if we push a vertex 2 times; the algorithm still takes O(n) time
                    // Because we can show each vertex won't be pushed more than twice, 
                    // we don't need to do any processing to make sure we don't push the same vertices 
                    dq.push_back(v);
                }
            }
        }
    }
    return dist[n];
}

int main(){
    int n,m;
    cin >> n >> m;
    vector<vector<int>> adj(n + 1);
    vector<vector<bool>> is_single(n + 1);
    for(int i = 0; i < m; i++){
        int u,v,b;
        cin >> u >> v >> b;
        adj[u].push_back(v);
        adj[v].push_back(u);
        is_single[u].push_back(b);
        is_single[v].push_back(b);
    }
    cout << bridges(n, adj, is_single) << endl;
}